src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,64 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-
-            // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            if (x is not null && y is not null)
             {
-                x = xParts[i];
-                y = yParts[i];
+                if (x!.Contains(ConfigurationPath.KeyDelimiter) && y!.Contains(ConfigurationPath.KeyDelimiter))
+                {
+                    string[] xParts = x!.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries);
+                    string[] yParts = y!.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries);
+
+                    // Compare each part until we get two parts that are not equal
+                    for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+                    {
+                        int result = compare(xParts[i], yParts[i]);
+                        if (result != 0)
+                        {
+                            // One of them is different
+                            return result;
+                        }
+                    }
+
+                    // If we get here, the common parts are equal.
+                    // If they are of the same length, then they are totally identical
+                    return xParts.Length - yParts.Length;
+                }
+                else
+                {
+                    return compare(x, y);
+                }
+            }
 
+            return x?.Length ?? 0 - y?.Length ?? 0;
+
+            static int compare(string? a, string? b)
```

Comment:
```suggestion
            static int Compare(string? a, string? b)
```
Local functions should also be named _PascalCase_.

Reply:
I was wondering about that. I really like _camelCase_ local functions, and indeed, [_camelCased_ private functions](https://dunnhq.com/posts/2018/06/26/247/).

Reply:
Here applies [C# Coding Style](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) §13
> We use PascalCasing for all method names, including local functions.

😉